### ISSUE TITLE
Binaries extraction occurring repeatedly in a single execution

### ISSUE THREAD

**Issue by hjort** [yandex-qatools/postgresql-embedded#20]
In short, I created a simple code with the following snippet using *postgresql-embedded*:

```
PostgresStarter<PostgresExecutable, PostgresProcess> runtime = PostgresStarter.getDefaultInstance();
final PostgresConfig config = PostgresConfig.defaultWithDbName("test");
PostgresExecutable exec = runtime.prepare(config);
PostgresProcess process = exec.start();
[...]
process.stop();
```

When running that code, the following message appears 4 times:

> Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz DONE

I've noticed that every time the method `PostgresProcess.runCmd()` is called internally, these two methods are called:

> de.flapdoodle.embed.process.runtime.Starter.prepare(CONFIG)
> de.flapdoodle.embed.process.extract.AbstractExtractor.extract(IDownloadConfig, File, FilesToExtract)

This happens when:
1. `PostgresExecutable exec = runtime.prepare(config)` is executed
2. `ru.yandex.qatools.embed.postgresql.PostgresProcess.onBeforeProcess(IRuntimeConfig)` calls `initdb`
3. `ru.yandex.qatools.embed.postgresql.PostgresProcess.onAfterProcessStart(ProcessControl, IRuntimeConfig)` calls `createdb`
4. `ru.yandex.qatools.embed.postgresql.PostgresProcess.sendStopToPostgresqlInstance()` calls `pg_ctl ... stop`

Once PostgreSQL binaries were extracted to a temporary directory, why do they need to be extracted again each time `PostgresProcess.runCmd()` is invoked?

Attached is the entire class code:
[PostgresEmbedded.pdf](https://github.com/yandex-qatools/postgresql-embedded/files/43115/PostgresEmbedded.pdf)

**Comment by smecsia** [comment]
This is actually the default behavior of how [de.flapdoodle.embed.process](https://github.com/flapdoodle-oss/de.flapdoodle.embed.process) works. It extracts the executable every time you launch it.  
You can avoid this by setting your own artifact store, like [here](
https://github.com/yandex-qatools/postgresql-embedded/blob/master/src/test/java/ru/yandex/qatools/embed/postgresql/TestPostgresStarter.java#L47-L55).

**Comment by hjort** [comment]
I tried setting my own artifact store such as the example, but it still extracts the binaries 4 times.

I even tried running the `ru.yandex.qatools.embed.postgresql.TestPostgresStarter` class, which executes the extraction several times:

```
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz extract pgsql/pgAdmin3/lib/libjpeg.so.9.1.0
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz DONE
2015-11-25 10:46:10,361 [main                          ] DEBUG ProcessControl                 - Detected pid: 30875
[...]
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz extract pgsql/pgAdmin3/lib/libjpeg.so.9.1.0
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz DONE
[...]
2015-11-25 10:46:10,361 [main                          ] DEBUG ProcessControl                 - Detected pid: 30875
2015-11-25 10:46:24,271 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@1ab6db4
2015-11-25 10:46:24,272 [main                          ] DEBUG ProcessControl                 - Detected pid: 30926
[...]
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz extract pgsql/pgAdmin3/lib/libjpeg.so.9.1.0
Extract /home/user/.embedpostgresql/postgresql-9.4.4-1-linux-binaries.tar.gz DONE
[...]
2015-11-25 10:46:52,447 [main                          ] DEBUG ProcessControl                 - Detected pid: 30947
2015-11-25 10:46:52,447 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@5c498e
2015-11-25 10:46:52,452 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@91a775
2015-11-25 10:46:52,511 [Thread-8                      ] DEBUG Files                          - could delete /tmp/postgresql-embed-9a3d9d7e-ee18-4d1c-a748-d264715b9625/pgsql/bin/createdb.pid
2015-11-25 10:46:52,512 [Thread-3                      ] DEBUG Files                          - could delete /tmp/postgresql-embed-9a3d9d7e-ee18-4d1c-a748-d264715b9625/pgsql/bin/initdb.pid
Nov 25, 2015 10:46:52 AM ru.yandex.qatools.embed.postgresql.PostgresProcess stopInternal
INFORMAÇÕES: trying to stop postgresql
2015-11-25 10:46:59,124 [Thread-1                      ] INFO  ProcessControl                 - execSuccess: false [kill, -2, 30926]
2015-11-25 10:46:59,125 [Thread-1                      ] DEBUG ProcessControl                 - Detected pid: 30974
2015-11-25 10:46:59,135 [Thread-1                      ] INFO  ProcessControl                 - execSuccess: false [kill, 30926]
2015-11-25 10:46:59,166 [Thread-1                      ] DEBUG Files                          - could delete /tmp/postgresql-embed-9a3d9d7e-ee18-4d1c-a748-d264715b9625/db-content-502823f8-eff5-4723-8c0a-8be41a6198a2
2015-11-25 10:46:59,166 [Thread-1                      ] DEBUG Files                          - could delete /tmp/postgresql-embed-9a3d9d7e-ee18-4d1c-a748-d264715b9625/pgsql/bin/postgres.pid
2015-11-25 10:46:59,284 [Thread-1                      ] DEBUG Files                          - could delete /tmp/postgresql-embed-9a3d9d7e-ee18-4d1c-a748-d264715b9625/pgsql
```

Should we avoid that by specifying other PostgreSQL commands (initdb, createdb, etc) along with `Command.Postgres` when creating the `IRuntimeConfig` object?

**Comment by smecsia** [comment]
Please try version 1.7 (I've just recently released it).

**Comment by hjort** [comment]
It worked successfully using version 1.7, thank you!

A question: do you have any code sample to avoid 1) downloading the binaries the first time and 2) extracting the ZIP file every execution? I mean, I'd like to retrieve and extract the binaries once.

```
Extract /home/user/.embedpostgresql/postgresql-9.1.19-1-linux-binaries.tar.gz extract pgsql/bin/vacuumlo
Extract /home/user/.embedpostgresql/postgresql-9.1.19-1-linux-binaries.tar.gz DONE
2015-11-26 13:47:49,986 [main                          ] DEBUG ProcessControl                 - Detected pid: 14272
2015-11-26 13:47:51,325 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@1783ddd
2015-11-26 13:47:51,326 [main                          ] DEBUG ProcessControl                 - Detected pid: 14317
2015-11-26 13:47:52,370 [main                          ] DEBUG ProcessControl                 - Detected pid: 14323
2015-11-26 13:47:52,371 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@158fa97
2015-11-26 13:47:53,371 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@5373c8
url = jdbc:postgresql://localhost:5131/test
MOVIES:
- movie
- filme
- pelic
Nov 26, 2015 1:47:53 PM ru.yandex.qatools.embed.postgresql.PostgresProcess stopInternal
INFORMAÇÕES: trying to stop postgresql
2015-11-26 13:47:54,480 [main                          ] DEBUG ProcessControl                 - Detected pid: 14331
2015-11-26 13:47:54,480 [main                          ] INFO  Executable                     - start ru.yandex.qatools.embed.postgresql.config.PostgresConfig@8e6de9
Nov 26, 2015 1:47:55 PM ru.yandex.qatools.embed.postgresql.PostgresProcess sendStopToPostgresqlInstance
INFORMAÇÕES: Cleaning up after the embedded process (removing /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8)...
2015-11-26 13:47:55,618 [main                          ] DEBUG Files                          - could delete /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8
Nov 26, 2015 1:47:55 PM ru.yandex.qatools.embed.postgresql.PostgresProcess deleteTempFiles
ADVERTÊNCIA: Could not delete temp db dir: /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8/db-content-4bb92935-20c3-4909-90bc-75eb145b29da
2015-11-26 13:47:55,618 [main                          ] WARN  AbstractProcess                - Could not delete pid file: /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8/pgsql/bin/postgres.pid
2015-11-26 13:47:55,619 [Thread-8                      ] WARN  AbstractProcess                - Could not delete pid file: /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8/pgsql/bin/createdb.pid
2015-11-26 13:47:55,620 [Thread-3                      ] WARN  AbstractProcess                - Could not delete pid file: /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8/pgsql/bin/initdb.pid
2015-11-26 13:47:55,620 [Thread-12                     ] WARN  AbstractProcess                - Could not delete pid file: /tmp/postgresql-embed-cb18b2b6-5c06-4c9a-8ccf-f7814a011ac8/pgsql/bin/pg_ctl.pid
```

**Comment by smecsia** [comment]
1) Binaries are being downloaded once for each version and then reside within $HOME/.embedpostgresql directory.
2) What do you mean by "extract the binaries once"? Do you want the binaries to be extracted during first run and then use the specific extracted directory during the following runs? Why do you need that?

**Comment by mblt** [comment]
> Why do you need that?

For instance, this can improve performance

**Comment by smecsia** [comment]
I mean the common use case for this library is to embed the PostgreSQL into your build process. Each build must be isolated and autonomous, so the intention is to use the clean environment. Besides that, extraction is a relatively quick process, I'm not sure if it is a big issue.

**Comment by mikko-apo** [comment]
+1 for extracting once, only if the extracted directory for the postgres version does not exist. If postgres does not generate any files to the extracted directory or modify any files there, it should be safe to assume that the postgres directory is ok to use between runs. This would speed up build processes.

Also, extraction messages produce unnecessary noice to the build log so it would be nice if there's output only if the extraction fails.

ps. Thanks for implementing the lib! It's going to hopefully help us in simplifying developer setup for the roughly 20 projects our customer is running here.

**Comment by mikko-apo** [comment]
btw, switching from 1.4 to 1.7 improved build time from 4 minutes to 3.5 for one project. postgres is started twice during the build: scala-tests and ui-tests are run sequantially in different processes.

**Comment by smecsia** [comment]
Guys. The latest released version (`1.11`) does not print all the extracted files in the log, so you can be sure that the single embedded postgres instance extracts files just once for each launch. 
Extraction is a quick process and should not slow down your tests much. I guess we can close this issue for now. Feel free to reopen if you have any new thoughts.
